tests/unit/test_query_language.py

```diff
@@ -381,3 +384,45 @@ def test_incompatible_duration_operations(lhs, op, rhs):
         else:
             assert False
         assert not result
+
+
+@pytest.mark.parametrize(
+    "fn_name",
+    (
+        {k for k, v in DateFunctions.__dict__.items() if callable(v)}
```

Comment:
:+1: I'm a fan of using introspection to make sure you can't add things without remembering to test them (the coverage check helps with this, but doesn't solve it completely).

A variant on what you've done here is to define the test parameters manually, but have a separate test that inspects the parameters and makes sure they cover everything e.g.
https://github.com/opensafely-core/databuilder/blob/8c50e843dbc86f62fa73a9a3717778823680611e/tests/unit/file_formats/test_csv.py#L43-L48

Sometimes, generating the test examples themselves by introspection can end up being full of conditional logic and hard to follow, and having manually curated examples and a separate exhaustiveness check can keep the tests simple at the cost of being a bit more laborious to construct.

I think the test here is simple enough to not be a problem, but worth bearing in mind if it starts getting more complicated.

Reply:
Cool - I did have some concerns re: the complexity of the derivation of the test parameters but figured it was just about on the right side of readable. I hadn't thought about using another test to check the params of a different test - that's a great idea!